### PR TITLE
feat(db): make the SDK generic over the store

### PR DESCRIPTION
**Pull request by joshuachp**
This PR make the SDK generic over the store and implements the MemoryStore, to keep the properties until the device is rebooted.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,13 +1,30 @@
 # Changelog
+
 All notable changes to this project will be documented in this file.
 
-The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
-and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project
+adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## Next
```
Comment (harlem88) [inline CHANGELOG.md:8]: `## Next` -> `## [1.0.0] - Unreleased`

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,13 +1,30 @@
...
+
+## Added
+
```
Comment (harlem88) [inline CHANGELOG.md:11]: Remove the new line

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,13 +1,30 @@
...
+- Add support for in-memory `MemoryStore` properties if no database is provided
+- Make `AstarteDeviceSdk` generic over the store type
+- Provide type aliases for `AstarteDeviceSdk` with `MemoryStore` and `SqliteStore`
+
+## Changed
+
```
Comment (harlem88) [inline CHANGELOG.md:17]: Remove the new line

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,13 +1,30 @@
...
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project
+adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## Next
+
+## Added
+
+- Add support for in-memory `MemoryStore` properties if no database is provided
+- Make `AstarteDeviceSdk` generic over the store type
+- Provide type aliases for `AstarteDeviceSdk` with `MemoryStore` and `SqliteStore`
```
Comment (harlem88) [inline CHANGELOG.md:14]: Add `.` at the end of line.
``` md
- Add support for in-memory `MemoryStore` properties if no database is provided.
- Make `AstarteDeviceSdk` generic over the store type.
- Provide type aliases for `AstarteDeviceSdk` with `MemoryStore` and `SqliteStore`.
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,13 +1,30 @@
...
+
+## Changed
+
+- Renamed the `AstarteSqliteDatabase` into `SqliteStore`
```
Comment (harlem88) [inline CHANGELOG.md:18]: Add `.` at the end of line.
```
- Renamed the `AstarteSqliteDatabase` into `SqliteStore`.
```

File: src/database/mod.rs [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+// NOTE: the 'static bound is required for the MSRV, in other version the error is not present
+#[async_trait]
+pub trait AstarteDatabase: Debug + Sync + 'static
+where
+    Self::Err: StdError,
+{
+    type Err;
+
+    /// Stores a property within the database.
+    async fn store_prop_impl(
```
Comment (harlem88) [inline src/database/mod.rs:47]: Why did you add `impl` functions? I think it's a little bit confusing

File: .gitattributes [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# This file is part of Astarte.
```
Comment (harlem88) [inline .gitattributes:1]: I think that these changes can be moved to a separate PR.

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: .gitattributes (hunk 7 above)
Comment (joshuachp) [reply .gitattributes:1]: This will fix an issue with the checks at compile time on Windows platforms.

Comment (joshuachp) [reply .gitattributes:1]: The binary cache was added because testing the PR took 7 minutes plus each time

File: src/utils.rs [hunk 8]
```diff
@@ -0,0 +1,112 @@
+// This file is part of Astarte.
```
[the remark this answers is not in the document]
Comment (joshuachp) [reply src/utils.rs:1]: Done

File: src/database/memory.rs [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+use super::{AstarteDatabase, StoredProp};
+use crate::{types::AstarteType, AstarteError};
+
+/// Data structure providing an implementation of an in memory Key Value Store.
+///
+/// Can be used by an Astarte device to store variables while the device is running.
+#[derive(Debug, Clone, Default)]
+pub struct MemoryStore {
+    // Tore the properties in memory
```
[the remark this answers is not in the document]
Comment (joshuachp) [reply src/database/memory.rs:35]: Done

File: src/database/memory.rs [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+        store
+            .entry(key)
+            .and_modify(|val| {
+                val.value = value.clone();
+                val.interface_major = interface_major;
+            })
+            .or_insert_with(|| Value {
+                value: value.clone(),
+                interface_major,
+            });
```
[the remark this answers is not in the document]
Comment (joshuachp) [reply src/database/memory.rs:73]: Done

File: src/database/mod.rs (hunk 6 above)
Comment (harlem88) [reply src/database/mod.rs:47]: Yes, I prefer the second option

File: examples/individual_properties/main.rs [hunk 11]
```diff
@@ -54,7 +59,7 @@ async fn get_name_for_sensor(device: &AstarteDeviceSdk, sensor_n: i32) -> Result
 }
 
 #[tokio::main]
-async fn main() -> Result<(), AstarteError> {
+async fn main() -> Result<(), Box<dyn StdError>> {
```
Comment (harlem88) [inline examples/individual_properties/main.rs:62]: Can you use the `astarte_device_sdk::error::Error`?
```suggestion
async fn main() -> Result<(), astarte_device_sdk::error::Error> {
```

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: examples/individual_properties/main.rs (hunk 11 above)
Comment (joshuachp) [reply examples/individual_properties/main.rs:62]: No because there are multiple different errors. We should use a library like Anyhow here, but an easier way without dependencies is to use Box<dyn StdError>

File: src/database/mod.rs (hunk 6 above)
Comment (joshuachp) [reply src/database/mod.rs]: I created a StorageWrapper for wrapping the error from the storage, with this I deleted the impl methods from the trait.

**Comment by joshuachp** [comment]
Closing this since it has been separated in #207, #210, #211, #212, #213, #214 and #215
